langages/php/les-arrays-php-sont-ils-des-arrays.md

```diff
@@ -0,0 +1,181 @@
+
+```php
+// An array can be created using the array() language construct. It takes any number of comma-separated key => value pairs as arguments.
+// A short array syntax exists which replaces array() with []. 
+
+array(
+    key  => value,
+    key2 => value2,
+    key3 => value3,
+    ...
+)
+```
+
+On parle bien ici ce paire de clés-valeurs : l'array php est donc bien un dictionnaire ?
+
+
+### Mais qu'en est-il de la syntaxe sans clés ?
+
+En effet il existe une seconde syntaxe possible où aucune clé n'est renseignée. Si on regarde la doc :
+
+```php
+// The key is optional. If it is not specified, PHP will use the increment of the largest previously used int key.
+<?php
+$array = array("foo", "bar", "hello", "world");
+var_dump($array);
+?>
+
+// The above example will output:
+array(4) {
+  [0]=>
+  string(3) "foo"
+  [1]=>
+  string(3) "bar"
+  [2]=>
+  string(5) "hello"
+  [3]=>
+  string(5) "world"
+}
+```
+
+De fait, certes on n'a mentionné aucune clé mais elles sont toujours là, *les clés sont seulement définie implicitement par le compilateur*.
+
+À partir de ce constat, la conclusion me semble assez évidente : les arrays de PHP sont en fait indéniablement des dictionnaires.
+
+
+## Qu'en est-il à l'usage ?
+
+Malgré cette conlusion, il me semblait intéressant d'ajouter un dernier point sur l'usage que l'on peut avoir des arrays PHP, en dehors se sa pure définition.
+
+En premier lieu et pour simplifier mon propos je vais introduire le terme de liste. Une liste en PHP est un array dont les clés sont constituées de nombres consécutifs de 0 à count($array)-1. Cela correspond à la définition fournie par la fonction standarde de PHP [array_is_list](https://www.php.net/manual/en/function.array-is-list.php). 
+
+Par extension et si on reprend les exemples de syntaxe au dessus, une liste est ce que founit PHP quand on ne mentionne aucune clé dans son tableau, soit donc le cas le plus éloigné syntaxiquement du dictionnaire.
+
+
+### Tests de l'array PHP
+
+Pour tester comme PHP traite les listes, j'ai testé le comportement de plusieurs fonctions qui insèrent ou enlèvent des valeurs pour voir comment les clés sont traitées. Les fonctions testée sont:
+
+- array_filter
+- array_shift
+- array_slice
+- array_splice
+- array_splice pour un ajout
+- array_unshift
+- shuffle (car il modifie l'ordre)
+
+Pour être sûr de ne rien louper je fais les tests avec 
+
+- un array avec clés explicites
+- une liste avec clés explicites
+- une liste avec clés explicites dans l'ordre inverse
+- une liste avec clés implicites.
+
+Mon script ressemble à ceci pour array_filter :
+
+```php
+<?php
+// $numbers = ["zero" => 0, "un" => 1, "deux" => 2];
+// $numbers = [0 => 0, 1 => 1, 2 => 2];
+// $numbers = [2 => 0, 1 => 1, 0 => 2];
+$numbers = [0, 1, 2];
+
+
+$test = array_filter($numbers, fn($value) => ($value % 2) !== 0);
+var_dump("Array filter");
+var_dump($test);
+```
+
```

Comment:
ça serait bien d'afficher les résultats du coup

Reply:
Sur ce point j'avais peur de trop rallonger l’article et c'est aussi pour ça que je n'ai mis qu'une version tronquée de mon script de test.

Si le résumer ne suffit pas, peut-être donner dans ce cas 2-3 exemples de codes avec les différentes nuances, si ça ne prolonge pas trop l'article ?

Reply:
oui, je suis sur que tu peux faire un exemple conçis qui ne prend pas 50 lignes